lectures/2-math-and-crypto.tex

```diff
@@ -0,0 +1,269 @@
+Athough, with such an approach, we might get a false positive, as you cannot state for sure that $n$ is prime. For example, consider $n = 15$ and $a = 4$.
+$4^{15-1} \equiv 1 \pmod{15}$, but $n = 15 = 3 \cdot 5$ is composite. To solve this issue, $a$ is picked many times, decreasing the chances of a false positive.
+The probability that a composite number is mistakenly called prime for $k$ iterations is $2^{-k}$ = $\dfrac{1}{2^k}$.
+
+There exists a problem with such an algorithm in the form of \textbf{Carmichael numbers}, which are numbers that are Fermat pseudoprime to all bases. To put it simply,
+no matter how many times you check whether the number is prime using this type of primality test, it will always stay positive, even though the number is
+composite. The good thing is that Carmichael numbers are pretty rare. The bad thing is that there are infinitely many of them. 
+
+Even though this algorithm is probabilistic (which does not guarantee the correctness of the output) and has a vulnerability in the form of \textit{Carmichael numbers},
+it runs with an asymptotic complexity $O(\log^3{n})$. This is much better for large numbers and is often used in cryptography. Here is a pseudocode implementation of this algorithm:
+
+\begin{lstlisting}[language=Python]
+# n = number to be tested for primality
+# k = number of times the test will be repeated
+def is_prime(n, k):
+    i = 1
+    while i <= k:
+        a = rand(2, n - 1)
+
+        if a^(n - 1) != 1 (mod n):
+            return False
+
+        i++
+
+    return True
+\end{lstlisting}
+
+Miller-Rabin primality test, is a more advanced form of Fermat primality test. The main difference is it is not vulnerable to \textit{Carmichael numbers},
+which makes it much better to use in practice.
+
+\subsubsection{Reed-Solomon codes}
+
+Reed-Solomon codes allows to restore lost or corrupted data, implement threshold secret sharing and is used in some ZK protocols.
+Given a vector of data $V$ a polynom $P$ is constructed using Lagrange interpolation.
+Polynomial with degree $n$ can be uniquely defined using $(n+1)$ unique points. Defining more points on the same polynomial add a redundancy,
+which can be used to restore the polynomial even if some points are missing.
+Common choices for a set of evaluation points include ${0, 1, 2, ..., n - 1}$.
+
+The error-correcting ability of a Reed-Solomon code is $n-k$, the measure of redundancy in the block. 
+If the locations of the error symbols are not known in advance, then a Reed-Solomon code can correct up to
+$n-k/2$ erroneous symbols, i.e., it can correct half as many errors as there are redundant symbols added to the block.
+
+\subsubsection{Schwartz-Zippel Lemma}
+
+\begin{lemma}
+Let $\F_p$ be a field. Let $f(x_1, x_2, ..., x_n)$ be a polynomial of total degree $d$. Suppose that $f$ is not the zero polynomial. Let $S$ be
```

Comment:
Also, theorem holds for any finite field (including extensions), so I believe this should be added as a mentioning